File: src/symbol_table.py
class SymbolTable:
    def __init__(self):
        self.scopes = [{}] 

    def enter_scope(self):
        self.scopes.append({})

    def exit_scope(self):
        if len(self.scopes) > 1:
            self.scopes.pop()

    def define(self, name, type_, value=None):
        current_scope = self.scopes[-1]
        current_scope[name] = {'type': type_, 'value': value}

    def lookup(self, name):
        for scope in reversed(self.scopes):
            if name in scope:
                return scope[name]
        return None

    def lookup_in_current_scope(self, name):
        return name in self.scopes[-1]

    def print_table(self):
        print("\n--- Symbol Table Status ---")
        for i, scope in enumerate(reversed(self.scopes)):
            level = len(self.scopes) - 1 - i
            print(f"Scope Level {level}:")
            for name, info in scope.items():
                print(f"  - {name} : {info}")
        print("---------------------------")

File: src/main/java/symboltable/SymbolTable.java
package symboltable;

import java.util.ArrayList;
import java.util.List;

public class SymbolTable {
    public static class Symbol {
        String name;
        String type;
        String scope;

        public Symbol(String name, String type, String scope) {
            this.name = name;
            this.type = type;
            this.scope = scope;
        }
    }

    private List<Symbol> symbols = new ArrayList<>();
    private String tableName;

    public SymbolTable(String tableName) {
        this.tableName = tableName;
    }

    public void define(String name, String type, String scope) {
        symbols.add(new Symbol(name, type, scope));
    }

    public void printTable() {
        System.out.println("\n=======================================================");
        System.out.println("   " + tableName + " SYMBOL TABLE SNAPSHOT");
        System.out.println("=======================================================");
        System.out.printf("| %-20s | %-15s | %-10s |\n", "NAME", "TYPE", "SCOPE");
        System.out.println("-------------------------------------------------------");
        for (Symbol s : symbols) {
            System.out.printf("| %-20s | %-15s | %-10s |\n", s.name, s.type, s.scope);
        }
        System.out.println("=======================================================\n");
    }
}